.cursor/rules/data-encoding.mdc

```diff
@@ -0,0 +1,292 @@
+---
+description: 
+globs: 
+alwaysApply: true
+---
+ ---
+description: Guidelines for encoding and decoding data in WAVS components
+globs: components/**/*
+---
+# Data Encoding and Decoding Guidelines
+
+Guidelines for handling data encoding and decoding in WAVS components.
+
+## Input Decoding
+
+1. Trigger Event Decoding:
+```rust
+fn decode_trigger_event(trigger_data: TriggerData) -> Result<(u64, Vec<u8>, Destination)> {
+    match trigger_data {
+        // Handle Ethereum contract events
+        TriggerData::EthContractEvent(TriggerDataEthContractEvent { log, .. }) => {
+            // Decode event using generated types
+            let event: solidity::NewTrigger = decode_event_log_data!(log)?;
+            let trigger_info = solidity::TriggerInfo::abi_decode(&event._triggerInfo, false)?;
+            Ok((trigger_info.triggerId, trigger_info.data.to_vec(), Destination::Ethereum))
+        }
+        // Handle raw CLI input
+        TriggerData::Raw(data) => Ok((0, data.clone(), Destination::CliOutput)),
+        _ => Err(anyhow::anyhow!("Unsupported trigger data type")),
+    }
+}
+```
+
+2. Raw Input Processing:
+```rust
+// Convert bytes to string
+let input = std::str::from_utf8(&req)
+    .map_err(|e| e.to_string())?;
+
+// Parse first character as hex digit
+let value = input
+    .chars()
+    .next()
+    .ok_or("Empty input")?
+    .to_digit(16)
+    .ok_or("Invalid hex digit")? as u64;
+```
+
+## Output Encoding
+
+1. Basic Output Encoding:
+```rust
+fn encode_trigger_output(trigger_id: u64, output: impl AsRef<[u8]>) -> Vec<u8> {
+    solidity::DataWithId {
+        triggerId: trigger_id,
+        data: output.as_ref().to_vec().into()
+    }.abi_encode()
+}
+```
+
+2. Destination-based Encoding:
+```rust
+let output = match dest {
+    // For Ethereum: Encode with trigger ID
+    Destination::Ethereum => Some(encode_trigger_output(trigger_id, &result)),
+    // For CLI: Return raw output
+    Destination::CliOutput => Some(result),
+};
+```
```

Comment:
More explanation could be used here. `Destination` was a type in the oracle example. I think it's better to explain and document the expected types for each output.

Reply:
saving this comment here for when I make an issue: 

> See my comment above about explaining destinations and outputs better so we know the encoding required for CLI output and EVM output.